lib/gollum-lib/file.rb

```diff
@@ -41,24 +44,32 @@ def path_compare(query, match_path, hyphened_tags, case_insensitive)
     # version - The String version ID to find.
     # try_on_disk - If true, try to return just a reference to a file
     #               that exists on the disk.
-    # global_match - If true, find a File matching path's filename, but not it's directory (so anywhere in the repo)
+    # global_match - If true, find a File matching path's filename, but not its directory (so anywhere in the repo)
     #
     # Returns a Gollum::File or nil if the file could not be found. Note
     # that if you specify try_on_disk=true, you may or may not get a file
     # for which on_disk? is actually true.
     def self.find(wiki, path, version, try_on_disk = false, global_match = false)
-      map = wiki.tree_map_for(version.to_s)
-
-      query_path = Pathname.new(::File.join(['/', wiki.page_file_dir, path].compact)).cleanpath.to_s
-      query_path.sub!(/^\/\//, '/') if Gem.win_platform? # On Windows, Pathname#cleanpath will leave double slashes at the start of a path intact, so sub them out.
+      query_path =  self.canonical_path(path, page_file_dir: wiki.page_file_dir)
+      path_segments = query_path.split('/')
+      filename = path_segments.last
+      dir = path_segments[0..-2].join('/')
 
-      begin
-        entry = map.detect do |entry|
-          path_match(query_path, entry, global_match, wiki.hyphened_tag_lookup, wiki.case_insensitive_tag_lookup)
+      if global_match && self.respond_to?(:global_find) # Only implemented for Gollum::Page
+        return self.global_find(wiki, version, query_path, try_on_disk) 
+      else
+        begin
+          root = wiki.commit_for(version.to_s)
+          return nil unless root
+          tree = dir.empty? ? root.tree : root.tree / dir
+          return nil unless tree
```

Comment:
Just as a sanity check: can `root.tree` ever be `nil`? Or can `root.tree / dir` be `nil`? Otherwise this check might not be needed.

Reply:
If `version` can only be a Commit (see other comment), then presumably it will never be `nil`. If Commit is/can be a String, `version` could be `nil`.

Reply:
It looks like `root.tree / dir` can be `nil`, when the user looks for a non-existent path. So best to keep the `nil` check.